docs/enhancements/fep-001-api-fleet-templating.md

```diff
@@ -0,0 +1,303 @@
+            cluster: …
+            namespace: factory-{{ .device.metadata.label[factory] }}
+            name: wifi-access-credentials
+            mountPath: /etc/network/…
+        - name: motd-update
+          configType: InlineConfigProviderSpec
+          inline:
+            ignition:
+              version: 3.4.0
+            storage:
+              files:
+                - contents:
+                    source: >-
+                        data:,{{ .device.name }}%0A
+                  mode: 422
+                  overwrite: true
+                  path: "/etc/acm-cluster-id"
+```
+
+The exposed values for use in templates is intentionally limited:
+
+* device.metadata.labels
+* device.metadata.name
+
+Making sure that the render of the device template is deterministic (i.e. not including timestamps or dates, etc..).
+
+#### Expected behavior
+
+When `fleet.spec.template` is updated, the fleet controller
+freezes the template, including non-expanded variables in the TemplateVersion. All existing devices
+are scanned to determine possible git/image/secret sources references by the `TemplateVersion` in the
+fleet. Those source references are captured in/alongside the `TemplateVersion` object. For non-versioned
+configuration sources (Vault secrets, K8s config maps, etc.), a snapshot of the configuration will be
+created and referenced under a (sortable) pseudo-version (e.g. `secret-$namespace-$name-$datetime-$hash`)
+
+Once the fleet controller has assigned the `fleet-controller/templateVersion` annotation to the fleet
+the rollout phase starts.
+
+When the fleet controller applies a `TemplateVersion` to a specific device through the Device API:
+* The template variables get rendered (i.e. the device's labels at that point in time get used)
+
+* An audit log entry is created for the device, capturing the rendered template and variables used
+  for expansion.
+
+* When later in time a device has labels updated (which could potentially result in template
+  evaluation changes) the device controller will queue the device to be re-evaluated and the
+  `TemplateVersion` re-rendered.
+
+#### Expected behavior on issues
+
+##### Device cannot be rendered from Template due to missing label
+When a device cannot be rendered for some reason (e.g. missing label), the fleet-controller
+will mark the device with a label `fleet-controller/failed-to-reconcile=true`, and add an annotation
+`fleet-controller/failed-to-reconcile-reason` with a human-readable reason. The fleet will include
+a status condition `DeviceFailedToReconcile`.
+
+Once all devices are corrected and can be reconciled again this status condition will be removed.
+
+#### Device cannot generate /device/{name}/rendered version due to missing resource
+When a device cannot render the template due to missing resources (e.g. git repo not found, secret not found),
+the device-controller will set a status condition on the device `MissingResource` with a human-readable reason.
```

Comment:
We already catch these "not found" errors when freezing the template, and in that case we set the Valid condition on the fleet to False.